### ISSUE TITLE
Elliptic Curve for Conch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8370]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)| @the0id reported|
|-|-|
|Trac ID|trac#8370|
|Type|enhancement|
|Created|2016-05-31 14:49:00Z|

A proposal for adding Elliptic Curve crypto to Conch.

This ticket supersedes ticket [[[[#8325](https://github.com/twisted/twisted/issues/8325)](https://github.com/twisted/twisted/issues/8325) | #8325]] as it requires the changes in that ticket, but also adds support for ECDSA.

Good news, Everyone!  I figured out what was going wrong with ECDSA.  When you have a signature r or s value with the [[| most significant bit on](http://crypto.stackexchange.com/questions/1795/how-can-i-convert-a-der-ecdsa-signature-to-asn-1)], you must prepend a null byte. 

~~NOTE: THIS IS NOT WORKING 100%.  Working on implementing Elliptic Curve has been a good way for me to learn about it (which I know is evident in some of my code), but I'm having some trouble getting ECDSA to work in all cases.~~ 

~~It works fine when using a Twisted SSH server and Conch, as well when using a OpenSSH server and Conch, but it only works about 40% of the time when using a Twisted SSH server and an OpenSSH client.~~

~~This part has me perplexed.  It gives a "bad format" error, but not specifically where.  I've tried looking for differences using Wireshark and various changes in the code, but have been unable to find out what is going wrong or where.  To make matters even more confusing is the fact that it works about 40% of the time.  I would expect it to work either 100% or 0% of the time.~~

~~So because I am at a loss of what could be happening, and that my time available to work on this is diminishing, I'm posting this hoping someone can finish it up and show me what's wrong.~~

Attachments:

* [_kex.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a4f/a4ffdbfa28cb353ce6106caf380ea9c5b44d4af5/5864b6ae3be40a647b4a69be6e0aa1de7cb09adf.patch) (4117 bytes) - added by the0id on 2016-05-31 14:50:06Z - Changes to _kex.py
* [keys.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a4f/a4ffdbfa28cb353ce6106caf380ea9c5b44d4af5/ba3c2fff8421ccc5a6d9b6f1bf5723f94ad9fca6.patch) (18089 bytes) - added by the0id on 2016-05-31 14:50:27Z - Changes to keys.py
* [transport.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a4f/a4ffdbfa28cb353ce6106caf380ea9c5b44d4af5/4f6281a84a17bd198b0ff65d6fc77532e40023af.patch) (12265 bytes) - added by the0id on 2016-05-31 14:50:44Z - Changes to transport.py
* [factory.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a4f/a4ffdbfa28cb353ce6106caf380ea9c5b44d4af5/232d1109932947a0ad54cdb60ba1704ce574e431.patch) (598 bytes) - added by the0id on 2016-05-31 14:51:03Z - Changes to factory.py
* [bundle.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a4f/a4ffdbfa28cb353ce6106caf380ea9c5b44d4af5/f8a0e18d47bc1a5c63a80cb03fbbb7b4803f434d.patch) (35069 bytes) - added by the0id on 2016-05-31 14:51:21Z - All of the above changes in one patch.

<details><summary>Searchable metadata</summary>

```
trac-id__8370 8370
type__enhancement enhancement
reporter__the0id the0id
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__conch conch
keywords__None None
time__1464706140208625 1464706140208625
changetime__1478266150162960 1478266150162960
version__None None
owner__the0id the0id
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_3"></a>|@rodrigc commented|
|-|-|

The process for submitting patches to the Twisted project has changed since you first
opened this ticket.  Since you've already opened this ticket in Trac,
can you read through the new documentation at:

https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch

and as per the instructions in that document, submit a GitHub pull request with your changes?

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_4"></a>|@rodrigc commented|
|-|-|

After you submit the GitHub pull request,
put a link to that pull request in this ticket.

Then add **review** back to the Keywords of this ticket so that this
goes back in the review queue.

Lot of rules....

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_5"></a>|@the0id commented|
|-|-|

No problem, I'm more of a fan of Git anyway. :)

I did have a question though.  As I was working on my Git pull request, I was wondering if I should go ahead and do a full pull request since I know the code isn't 100% working right now.

It probably shouldn't be pushed into the main branch until the ECDSA issue is resolved, but if submitting a pull request is the best way to get someone to look at the code and fix it, then I can do that.  On top of that it appears there are some python 3 specific errors that I'm working to fix.

As of right now the code is in my [[| fork](https://github.com/the0id/twisted/tree/8370-the0id-sshEC)], but I'm going to hold off completing a pull request for now.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc commented|
|-|-|

That's fine, your branch had some errors, so the code needs more work, anyways.
File another PR whenever you feel you are ready.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

the0id - if your code is complete enough to have a ticket in review, it is by definition complete enough for a pull request :).

rather than pulling this out of review, I'm going to open a PR for you against that branch, so we can get CI started and have more expert people give you their help.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

PR at https://github.com/twisted/twisted/pull/387

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

BTW, please set your author email.  https://github.com/twisted/twisted/pull/387/commits/3edf6d4e1dcc493c03690bcd245df45e8dda41a5 for example shows a greyed-out octocat for your picture.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Smaller pull requests are better, so trying to implement _all_ the EC algorithms in one go seems like a bad idea, especially if you are struggling with unreliable bits of the implementation.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

OK, let's call it a full review.  Clearly this needs some more work; the main thing I think it needs is to be shortened as much as possible to be one useful addition, which means just one EC algorithm at first.  I put a bunch of other inline comments, including a potential clue as to its unreliability, on the github PR.  Please address those and resubmit for review (re-attach the `review` keyword here).

Thank you very much for contributing to Conch - this is a major feature deficiency that I would very much like to integrate, so I do hope you'll continue working on it!

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_12"></a>|@the0id commented|
|-|-|

Sure, I can roll back the ECDSA parts of the code and just focus on the ECDH part.

I'm thinking it'd make sense to just comment out, and not remove, the ECDSA parts since I'd imagine they'd just get put back in for the next step.

I'll push my changes once I get that done.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_13"></a>|@the0id commented|
|-|-|

On second thought I went ahead and removed the ECDSA lines instead of commenting them out to make the code a little cleaner.

The changes have been pushed to Git and I made a few comments there.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Thanks again for contributing this.  It looks like the CI is still failing; my inline comments on github constitute a full review at this point.  When you're done addressing them please resubmit.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_15"></a>|@the0id commented|
|-|-|

Not a problem. This is the first time I've been able to contribute to a community (you're shocked, I know), so I'm happy to be able to finally do so.  Thanks for working with me to get this through.

As mentioned on Github, I see a message that the CI is failing, but I can't find any details.  The only thing that responds in the CI box is a link to Appveyor's homepage.

I've uploaded the latest changes, including a merge of the latest trunk, so hopefully that will fix everything that's needed.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_16"></a>|@the0id commented|
|-|-|

Fixed the conflicts.  I removed the branch, merged my trunk with the most recent Twisted trunk, then recreated the branch.

The new pull request is at https://github.com/twisted/twisted/pull/432

The build fails, but for one expected reason and one unknown as I talk about in the comments there.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_17"></a>|@the0id commented|
|-|-|

I've made the needed changes and it looks like the build is succeeding now, so I'm marking this for review again (as well as the other ticket 8325 because they seem to have overlapped each other).

It may make things more confusing but I'm adding the ECDSA portion of the code into another branch https://github.com/the0id/twisted/tree/8370-the0id-sshECDSA

This still has the problems mentioned above that it works fine with the exception of using an OpenSSH host with a Conch server still only works 40% of the time.  Because of this I'm not going to create a PR, but I wanted the code out there, and maybe someone can see what I'm missing.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_18"></a>|@the0id commented|
|-|-|

As an added bonus, I've included a version with Curve25519 support.

Because Cryptography doesn't support Curve25519, I had to use two other libraries: [[| pynacl](https://github.com/pyca/pynacl)], and [[| curve25519-donna](https://github.com/agl/curve25519-donna)].

This is why I created a separate branch, and won't be issuing a PR.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Don't worry about the confusion!  We want each individual ticket to be merged with the smallest change to achieve the necessary functionality, mainly so we can be confident in the quality of the code review.  But you should by all means continue doing lots of work in _other_ branches in parallel, since you might discover useful information that can inform the design of this one.  Don't ever feel like you just need to stop writing code because something is awaiting a review; just try to put the code in other branches :).

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_20"></a>|@the0id commented|
|-|-|

Good, glad to see we're on the same page. :)

And to help clear things up, each of the branches build on each other, incorporating code from the one before.

sshECH-redux -> sshECDSA -> sshEC-plus-Curve25519

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

Thanks again for continuing with this work!  It looks like we've still got a little ways to go though.

The first and foremost problem is that your branch doesn't have any tests.  I can't believe I didn't make this as the first comment on the last review - I got lost in the crypto details which I find fairly mystifying, I suppose :).

You're adding new functionality, which means it should implicitly come along with tests that verify that functionality works.  In fact, when I typically do a review, I want to look at the tests first, because they express the intent of the changes, and combining test coverage information with the implementation of the tests, I can look very little at the implementation itself, because if the tests are right and the whole implementation is covered by the tests then of course the implementation has to be right :).

You should probably be looking in `twisted.conch.test.test_transport.ServerSSHTransportTests` which tests other kinds of key exchange.

I also put some inline comments (which I think a previous instance of might have gotten swallowed by your changing the author of the commits) on the github PR, mostly about coding standard stuff and exposing public API that we may not want to support in the future.  (If we do expose public API we need full documentation coverage of it, which you probably also don't want to write :)).

As far as code coverage goes, codecov.io is a bit unreliable, so if you look at all the coverage data you may find unrelated fluctuations. But when you're looking at your patch itself, generally the reporting is fairly accurate.  The most interesting part for you is to click on the link after "last update", which currently points [here](https://codecov.io/gh/twisted/twisted/compare/30ef07f4479c68c8f6b09e8630b862d3a319fee2...6878a553fd95d5c42e2698f2228c08414a59f6df?src=pr), and then hit 'x' on your keyboard to jump to the first un-covered or partial line, which tells you what you need to write tests for.  ("?" will show you other keyboard shortcuts).

Thanks again; I look forward to re-reviewing.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_22"></a>|@the0id commented|
|-|-|

Yeah, I didn't do the unit tests because while I'm a big supporter of them, my time to work on them is very limited, and my main focus of my efforts is to get something that works for my project.

So yeah, I'm that guy. :)

I did push up some tests and while it helped the coverage check, it's not as much as I expected, and there are some big tests missing because I couldn't figure out how to get them to work.

I haven't written unit tests for this framework before so I'm using other tests as reference, but I'm getting confused by some of the behaviors.

For example assertKexInitResponseForDH checks to see if self.proto.x[5:] is 64 bytes of \x99.  To me it looks like x should be a randomly generated string of bytes, and so I'm surprised this check passes.

I'm sure there's stuff going on behind the scenes, but I don't know if I have the time to dig back there and figure out what's going on.

So if it helps, I can meet half way and push up my broken tests.  Most likely they're mostly there, just need a few adjustments.

As for the public API, you're right. Like any bad developer I hate writing documentation, and will avoid it at any cost. :)

But I responded to the inline comments about why I kept some of them public.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_23"></a>|@adiroiban commented|
|-|-|

See my comments from [#8325](https://github.com/twisted/twisted/issues/8325)

I think that we should first look at adding EC support to t.conch.ssh.key.Key , then look at handling the keys with scripts/ckeygen , then have ECDC ... then ECDSA

I am removing this from the review queue as there is noting to review for now.

Please push the broken tests. Maybe someone else can continue working on them... should be better than nothing.

Thanks for the contribution!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

(I agree with adiroiban and to that end I have reviewed [#8798](https://github.com/twisted/twisted/issues/8798))

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_26"></a>|@the0id commented|
|-|-|

The PR is building as I type this, but I think I should have the code coverage taken care of now.

Turns out it wasn't as much work as I thought since one of the tests wasn't working as I thought, and once that was fixed I was able to copy it for the other section.

As for acabhishek942's PR, handle it how you'd like. I understand the desire to check in as small of chunks as possible, but the differences between our two PR's is pretty trivial in my opinion.

But it's your project, and I'm not being passive aggressive when I say, handle it as you wish. :)

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_27"></a>|@the0id set status to `closed`|
|-|-|

I'm closing this ticket for reason expressed in ticket [[[[#8325](https://github.com/twisted/twisted/issues/8325)](https://github.com/twisted/twisted/issues/8325) | #8325 ]].  This should help clean things up a little.
